lib/scholar/cluster/hierarchical.ex

```diff
@@ -0,0 +1,483 @@
+    {clades[perm], diss[perm], sizes[perm]}
+  end
+
+  defnp merge_clades(clades, count, pointers, pairwise, diss, sizes, links, n, update_fun) do
+    {{clades, count, pointers, pairwise, diss, sizes}, _} =
+      while {{clades, count, pointers, pairwise, diss, sizes}, links},
+            i <- 0..(Nx.size(links) - 1) do
+        # i < j because of how links is formed.
+        # i will become the new clade index and we "infinity-out" j.
+        j = links[i]
+
+        if j == n do
+          {{clades, count, pointers, pairwise, diss, sizes}, links}
+        else
+          # Clades a and b (i and j of pairwise) are being merged into c.
+          a = find_clade(pointers, i)
+          b = find_clade(pointers, j)
+          c = count + n
+
+          # Update clades
+          new_clade = Nx.stack([a, b]) |> Nx.sort() |> Nx.new_axis(0)
+          clades = Nx.put_slice(clades, [count, 0], new_clade)
+
+          # Update sizes
+          sc = sizes[i] + sizes[j]
+          sizes = Nx.indexed_put(sizes, Nx.stack([i, c]) |> Nx.new_axis(-1), Nx.stack([sc, sc]))
+
+          # Update dissimilarities
+          diss = Nx.indexed_put(diss, Nx.stack([count]), pairwise[i][j])
+
+          # Update pairwise
+          {pairwise, _} =
+            while {pairwise, {x = i, y = j, sa = sizes[i], sb = sizes[j], sc = sc}},
+                  z <- 0..(n - 1) do
+              if z == x or z == y or Nx.is_infinity(pairwise[[0, z]]) do
+                {pairwise, {x, y, sa, sb, sc}}
+              else
+                dac = pairwise[[x, z]]
+                dbc = pairwise[[y, z]]
+                dab = pairwise[[x, y]]
+                update = update_fun.(dac, dbc, dab, sa, sb, sc)
+
+                pairwise =
+                  pairwise
+                  |> Nx.indexed_put(Nx.stack([x, z]), update)
+                  |> Nx.indexed_put(Nx.stack([z, x]), update)
+                  |> Nx.indexed_put(Nx.stack([y, z]), update)
+                  |> Nx.indexed_put(Nx.stack([z, y]), update)
+
+                {pairwise, {x, y, sa, sb, sc}}
+              end
+            end
+
+          infinities = Nx.take_diagonal(pairwise)
+
+          pairwise =
+            pairwise
+            |> Nx.put_slice([j, 0], Nx.reshape(infinities, {1, n}))
+            |> Nx.put_slice([0, j], Nx.reshape(infinities, {n, 1}))
+
+          # Update pointers
+          indices = [i, j] |> Nx.stack() |> Nx.new_axis(-1)
+          pointers = Nx.indexed_put(pointers, indices, Nx.stack([c, c]))
+
+          {{clades, count + 1, pointers, pairwise, diss, sizes}, links}
+        end
+      end
+
+    {clades, count, pointers, pairwise, diss, sizes}
+  end
+
+  defnp find_clade(pointers, i) do
+    {i, _, _} =
+      while {_current = i, next = pointers[i], pointers}, next != -1 do
+        {next, pointers[next], pointers}
+      end
+
+    i
+  end
+
+  # Dissimilarity functions
+
+  defnp pairwise_euclidean(%Nx.Tensor{} = x) do
+    Scholar.Metrics.Distance.pairwise_euclidean(x, x)
```

Comment:
```suggestion
    Scholar.Metrics.Distance.pairwise_euclidean(x)
```

Reply:
Oh in that case we can just get rid of that `defnp` entirely. We'll point `dissimilarity_fun` at it directly up in `fit/2`.

Reply:
[885c5b5](https://github.com/elixir-nx/scholar/pull/187/commits/885c5b5bd33627aea53aa283836ee1d0fca189f2)